platforms/jvm/testing-jvm-infrastructure/src/main/java/org/gradle/api/internal/tasks/testing/junitplatform/JUnitPlatformTestExecutionListener.java

```diff
@@ -328,11 +328,33 @@ private TestDescriptorInternal createSyntheticTestDescriptorForContainer(TestIde
 
     private TestDescriptorInternal createTestDescriptor(TestIdentifier test, String name, String displayName) {
         TestDescriptorInternal parentDescriptor = findTestParentDescriptor(test);
-        String className = parentDescriptor == null ? JUnitSupport.UNKNOWN_CLASS : parentDescriptor.getName();
-        String classDisplayName = parentDescriptor == null ? JUnitSupport.UNKNOWN_CLASS : parentDescriptor.getClassDisplayName();
+        String className = createUnknownClassName(test, parentDescriptor);
+        String classDisplayName = createUnknownClassDisplayName(test, parentDescriptor);
         return new DefaultTestDescriptor(idGenerator.generateId(), className, name, classDisplayName, displayName);
     }
 
+    private String createUnknownClassName(TestIdentifier node, @Nullable TestDescriptorInternal parentDescriptor) {
+        return createUnknownName(node, parentDescriptor, TestDescriptorInternal::getName);
+    }
+
+    private String createUnknownClassDisplayName(TestIdentifier node, @Nullable TestDescriptorInternal parentDescriptor) {
+        return createUnknownName(node, parentDescriptor, TestDescriptorInternal::getClassDisplayName);
+    }
+
+    private String createUnknownName(TestIdentifier node, @Nullable TestDescriptorInternal parentDescriptor, Function<TestDescriptorInternal, @Nullable String> nameGetter) {
```

Comment:
:x: I think this has a weird name, now that I read it more. Shouldn't these methods be called something like `determineClassName`, `determineClassDisplayName`, and `determineName`? They're called unconditionally, not only when we don't know the name.